datacontract/engines/fastjsonschema/check_jsonschema.py

```diff
@@ -1,31 +1,112 @@
 import json
 import logging
 import os
+import threading
+from typing import List
 
 import fastjsonschema
+from fastjsonschema import JsonSchemaValueException
 
 from datacontract.engines.fastjsonschema.s3.s3_read_files import yield_s3_files
 from datacontract.export.jsonschema_converter import to_jsonschema
 from datacontract.model.data_contract_specification import DataContractSpecification, Server
 from datacontract.model.exceptions import DataContractException
 from datacontract.model.run import Check, Run
 
+# Thread-safe cache for primaryKey fields.
+_primary_key_cache = {}
+_cache_lock = threading.Lock()
 
-def validate_json_stream(model_name, validate, json_stream):
-    try:
-        logging.info("Validating JSON")
-        for json_obj in json_stream:
+
+def get_primary_key_field(schema: dict, model_name: str) -> str:
+    # Check cache first.
+    with _cache_lock:
+        if model_name in _primary_key_cache:
+            return _primary_key_cache[model_name]
+
+    # Find primaryKey field.
+    fields = schema.get("properties", {})
+    primary_key_field = None
+    for field_name, attributes in fields.items():
+        if attributes.get("primaryKey", False):
+            primary_key_field = field_name
+            break
+
+    # Cache the result.
+    with _cache_lock:
+        _primary_key_cache[model_name] = primary_key_field
+    return primary_key_field
+
+
+def get_primary_key_value(schema: dict, model_name: str, json_object: dict) -> str:
+    # Get the `primaryKey` field.
+    primary_key_field = get_primary_key_field(schema, model_name)
+    logging.error(f"######## primary_key_field: '{primary_key_field}'")
+    if not primary_key_field:
+        return None
+
+    # Return the value of the `primaryKey` field in the JSON object.
+    value = json_object.get(primary_key_field, None)
+    logging.error(f"######## value: '{value}'")
+    return value
+
+
+def process_exceptions(run, exceptions: List[DataContractException]):
+    if not exceptions:
+        return
+
+    # Define the maximum number of errors to process (can be adjusted by defining an ENV variable).
+    error_limit = int(os.getenv("DATACONTRACT_MAX_ERRORS", 500))
+
+    # Calculate the effective limit to avoid index out of range
+    limit = min(len(exceptions), error_limit)
+
+    # Add all exceptions up to the limit - 1 to `run.checks`.
+    run.checks.extend(
+        [
+            Check(
+                type=exception.type,
+                name=exception.name,
+                result=exception.result,
+                reason=exception.reason,
+                model=exception.model,
+                engine=exception.engine,
+                message=exception.message,
+            )
```

Comment:
Since we use a property `message` here, I suggest to add the property to the Check class as Optional[str].